core/templates/dev/head/components/state-editor/state-editor-properties-services/state-property.service.ts

```diff
@@ -15,75 +15,103 @@
-      // behavior is to always return true.
-      _isValid: function(value) {
-        return true;
-      },
-      // Updates the memento to the displayed value.
-      saveDisplayedValue: function() {
-        if (this.setterMethodKey === null) {
-          throw 'State property setter method key cannot be null.';
-        }
-
-        this.displayed = this._normalize(this.displayed);
-        if (!this._isValid(this.displayed) || !this.hasChanged()) {
-          this.restoreFromMemento();
-          return;
-        }
-
-        if (angular.equals(this.displayed, this.savedMemento)) {
-          return;
-        }
-
-        AlertsService.clearWarnings();
-
-        this.savedMemento = angular.copy(this.displayed);
-      },
-      // Reverts the displayed value to the saved memento.
-      restoreFromMemento: function() {
-        this.displayed = angular.copy(this.savedMemento);
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { AlertsService } from 'services/alerts.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class StatePropertyService {
+  // The name of the setter method in ExplorationStatesService for this
+  // property. THIS MUST BE SPECIFIED BY SUBCLASSES.
+  setterMethodKey: string;
+  displayed: any;
+  stateName: string;
+  savedMemento: any;
+
+  constructor(private alertsService: AlertsService) {
+    this.setterMethodKey = null;
+  }
+
+  private isEquivalent(a: any, b: any): boolean {
+    // Create arrays of property names.
+    var aProps = Object.getOwnPropertyNames(a);
+    var bProps = Object.getOwnPropertyNames(b);
+    if (aProps.length !== bProps.length) {
+      return false;
+    }
+    for (var i = 0; i < aProps.length; i++) {
+      var propName = aProps[i];
+      if (a[propName] !== b[propName]) {
+        return false;
       }
-    };
+    }
+    return true;
+  }
+
+  init(stateName: string, value: any): void {
+    if (this.setterMethodKey === null) {
+      throw 'State property setter method key cannot be null.';
+    }
+    // The name of the state.
+    this.stateName = stateName;
+    // The current value of the property (which may not have been saved to
+    // the frontend yet). In general, this will be bound directly to the UI.
+    this.displayed = cloneDeep(value);
+    // The previous (saved-in-the-frontend) value of the property. Here,
+    // 'saved' means that this is the latest value of the property as
+    // determined by the frontend change list.
+    this.savedMemento = cloneDeep(value);
+  }
+
+  // Returns whether the current value has changed from the memento.
+  hasChanged(): boolean {
+    return !this.isEquivalent(this.savedMemento, this.displayed);
+  }
+
+  // Transforms the given value into a normalized form. THIS CAN BE
+  // OVERRIDDEN BY SUBCLASSES. The default behavior is to do nothing.
+  _normalize(value: any): any {
```

Comment:
Is this really any? If yes, it's fine otherwise add a TODO(#7165)

Reply:
Done.